igvm/vm.py

```diff
@@ -494,28 +501,23 @@ def run_puppet(self, clear_cert, tx):
 
         self.block_autostart()
 
-        if tx:
-            tx.on_rollback(
-                'Kill puppet',
-                self.run,
-                'pkill -9 -f "/usr/bin/puppet agent -v --fqdn={}"'
-                .format(self.fqdn)
-            )
-            self.run(
-                '/usr/bin/puppet agent -v --fqdn={}'
-                ' --server {} --ca_server {} --no-report'
-                ' --waitforcert=60 --onetime --no-daemonize'
-                ' --skip_tags=chroot_unsafe'
-                ' && touch /tmp/puppet_success'
-                ' | tee {} ;'
-                ' test -f /tmp/puppet_success'
-                .format(
-                    self.fqdn,
-                    self.dataset_obj['puppet_master'],
-                    self.dataset_obj['puppet_ca'],
-                    '/var/log/puppetrun_igvm',
-                )
+        self.run(
+            '/usr/bin/puppet agent -v --fqdn={}'
+            ' --server {} --ca_server {} --no-report'
+            ' --waitforcert=60 --onetime --no-daemonize'
+            ' --skip_tags=chroot_unsafe'
+            ' && touch /tmp/puppet_success'
+            ' | tee {} ;'
+            ' test -f /tmp/puppet_success'
+            .format(
+                self.fqdn,
+                self.dataset_obj['puppet_master'],
+                self.dataset_obj['puppet_ca'],
+                '/var/log/puppetrun_igvm',
             )
+        )
+        if tx:
+            tx.on_rollback('Kill puppet', self.kill_puppet)
```

Comment:
Wasn't it correct to register this before running `puppet`?  We want to kill it if it fails to complete the run.